src/uct/ib/rc/accel/rc_mlx5_ep.c

```diff
@@ -228,8 +228,126 @@ ucs_status_t uct_rc_mlx5_ep_get_zcopy(uct_ep_h tl_ep, const uct_iov_t *iov, size
 
-ucs_status_t uct_rc_mlx5_ep_am_short(uct_ep_h tl_ep, uint8_t id, uint64_t hdr,
-                                     const void *payload, unsigned length)
+#if HAVE_IBV_EXP_DM
+typedef struct uct_rc_mlx5_dm_pack_data {
+    uint64_t   hdr;
+    size_t     length;
+    const void *payload;
+} uct_rc_mlx5_dm_pack_data_t;
+
+typedef union uct_rc_mlx5_dm_copy_data {
+    uct_rc_am_short_hdr_t am_hdr;
+    struct ibv_exp_tmh    tm_hdr;
+    uint64_t              out[2];
+    char                  in[sizeof(uint64_t) * 2];
+} UCS_S_PACKED uct_rc_mlx5_dm_copy_data_t;
+
+static size_t uct_rc_mlx5_ep_am_short_dm_pack(void *dest, void *arg)
+{
+    uct_rc_mlx5_dm_pack_data_t *data = (uct_rc_mlx5_dm_pack_data_t*)arg;
+    uint64_t *dst = dest;
+    *dst = data->hdr;
+    memcpy(dst + 1, data->payload, data->length);
+    return sizeof(data->hdr) + data->length;
+}
+
+/* DM memory should be written by 8 bytes (int64) to eliminate
+ * processor cache issues. to make this used uct_rc_mlx5_dm_copy_data_t
+ * structure where first hdr_len bytes are filled by message header
+ * and tail is filled by head of message */
+static void UCS_F_ALWAYS_INLINE
+uct_rc_mlx5_ep_copy_to_dm(uct_rc_mlx5_dm_copy_data_t *cache, size_t hdr_len,
+                          const void *payload, size_t length, void *dm)
+{
+    size_t len          = sizeof(*cache) - hdr_len;
+    const uint64_t* src = payload + len;
+    uint64_t* dst       = dm;
+
+    ucs_assert(sizeof(*cache) >= hdr_len);
+
+    /* copy head of payload to tail of cache */
+    memcpy(cache->in + hdr_len, payload, ucs_min(len, length));
+
+    UCS_STATIC_ASSERT(sizeof(*cache) == sizeof(cache->out));
+    UCS_STATIC_ASSERT(sizeof(cache->in) == sizeof(cache->out));
+
+    /* atomically by 8 bytes copy data to DM */
+    *(dst++) = cache->out[0];
+    *(dst++) = cache->out[1];
+
+    for (; len < length; len += sizeof(*dst)) {
+        *(dst++) = *(src++);
+    }
+}
+
+
+ucs_status_t
+uct_rc_mlx5_ep_am_short_dm(uct_ep_h tl_ep, uint8_t id, uint64_t hdr,
+                           const void *payload, unsigned length)
+{
+    uct_rc_iface_t *iface = ucs_derived_of(tl_ep->iface, uct_rc_iface_t);
+    uct_rc_mlx5_ep_t *ep  = ucs_derived_of(tl_ep, uct_rc_mlx5_ep_t);
+    uct_rc_iface_send_desc_t *desc;
+    size_t len;
+    uct_rc_mlx5_dm_pack_data_t pack_data;
+    uct_rc_mlx5_dm_copy_data_t cache;
+
+    if (sizeof(uct_rc_am_short_hdr_t) + length < UCT_IB_MLX5_AM_MAX_SHORT(0)) {
+        return uct_rc_mlx5_ep_am_short_inline(tl_ep, id, hdr, payload, length);
+    } else {
+        UCT_CHECK_LENGTH(length + sizeof(uct_rc_hdr_t), 0,
+                         iface->dm.seg_len, "am_short");
+        UCT_CHECK_AM_ID(id);
+        UCT_RC_CHECK_RES(iface, &ep->super);
+        UCT_RC_CHECK_FC(iface, &ep->super, id);
+
+        desc = ucs_mpool_get_inline(&iface->dm.dm->mp);
+        if (ucs_unlikely((desc) == NULL)) {
+            /* in case if no resources available - fallback to bcopy */
+            pack_data.hdr = hdr;
```

Comment:
maybe unite it with tag_short fallback somehow?

Reply:
may be, but there are 2 calls only to unify, no reason for now

Reply:
but the fallback related code is quite big. Would be much cleaner just to call some fallback func

Reply:
yes, but there is only call `uct_rc_mlx5_txqp_bcopy_post` is same, all other stuff is unique for each routine